packages/basic-crawler/src/internals/basic-crawler.ts

```diff
@@ -1043,6 +1047,44 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
         return kvs.getAutoSavedValue<State>(BasicCrawler.CRAWLEE_STATE_KEY, defaultValue);
     }
 
+    protected get pendingRequestCountApproximation(): number {
+        let result = 0;
+
+        if (this.requestQueue) {
+            result += this.requestQueue.getPendingCount();
+        }
+        if (this.requestList) {
+            result += this.requestList.length() - this.requestList.handledCount();
+        }
+
+        return result;
+    }
+
+    protected calculateEnqueuedRequestLimit(explicitLimit?: number): number | undefined {
+        if (this.maxRequestsPerCrawl === undefined) {
+            return explicitLimit;
+        }
+
+        const limit = Math.max(
+            0,
+            this.maxRequestsPerCrawl - this.handledRequestsCount - this.pendingRequestCountApproximation,
+        );
+
+        return Math.min(limit, explicitLimit ?? Infinity);
+    }
+
+    protected async handleSkippedRequest(options: Parameters<SkippedRequestCallback>[0]): Promise<void> {
+        if (options.reason === 'limit' && this.shouldLogMaxEnqueuedRequestsExceeded) {
+            this.log.info(
+                'The number of requests enqueued by the crawler reached the maxRequestsPerCrawl limit of ' +
+                    `${this.maxRequestsPerCrawl} requests and no further requests will be added.`,
+            );
+            this.shouldLogMaxEnqueuedRequestsExceeded = false;
+        }
+
+        this.onSkippedRequest?.(options);
```

Comment:
I see that `this.onSkippedRequest` is potentially `Awaitable<>`, should we `await` this here? I suppose you want `handleSkippedRequest` async, right? (`Promise.all` below)